### PR TITLE
HTTP server refactor

### LINKED ISSUE #148: Feature request: Fix HTTP route for favicon.ico
1. What's the feature?

2. Extra information?

### PR DESCRIPTION
**Pull request by qdm12**
- Fixes #148
- Paves way for #75 
- Unblocks #159 
- Also makes the call to `GET /update` blocking until all updates are done

I skipped adding unit tests as I would like to get moving with an API server and a client side UI (react) which would change the routes quite a bit (in a subsequent PR). Also sorry for the delay @fredericrous was caught up in other repos essentially 😄

### CODE REVIEW

**Comment by fredericrous** [comment]
I'll test that now

**Comment by fredericrous** [comment]
The server hangs

```
ddns-updater on  http-server [$!?] via 🐹 v1.15.7 ➜  LISTENING_PORT=4242 ./app                                                                                17:19:45
=========================================
=========================================
========= DDNS Universal Updater ========
=========================================
=== Made with ❤️  by github.com/qdm12 ====
=========================================

Running version unknown built on an unknown date (commit unknown)


🔧  Need help? https://github.com/qdm12/ddns-updater/issues/new
💻  Email? quentin.mcgaw@gmail.com
☕  Slack? Join from the Slack button on Github
💸  Help me? https://github.com/sponsors/qdm12
2021-02-07T17:19:46.684Z        INFO    Found 2 settings to update records
2021-02-07T17:19:46.988Z        INFO    Reading history from database: domain mydomain.com host @
2021-02-07T17:19:46.988Z        INFO    Reading history from database: domain mydomain.com host *
```

It should continue and display the following lines but doesn't

```
2021-02-07T16:29:54.670Z        INFO    healthcheck server: listening on 127.0.0.1:9999
2021-02-07T16:29:54.670Z        INFO    http server: listening on 0.0.0.0:8000
```

The issue seems to come from this line, or from the update Necessary function

`r.forceResult <- r.updateNecessary(ctx)`

If I print something before this line it displays, but after it doesn't.

From what I understand of Go, the channel should be closed in the necessary function and the error should be returned as a string instead of a stream, am I right?

File: cmd/updater/main.go [hunk 1]
```diff
@@ -150,9 +150,9 @@ func _main(ctx context.Context, timeNow func() time.Time) int {
 	runner := update.NewRunner(db, updater, ipGetter, p.cooldown, logger, timeNow)
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	forceUpdate := make(chan struct{})
-	go runner.Run(ctx, p.period, forceUpdate)
-	forceUpdate <- struct{}{}
+
+	go runner.Run(ctx, p.period)
+	runner.ForceUpdate()
```
Comment (fredericrous) [inline cmd/updater/main.go:155]: this hangs, should we replace this by something like `go runner.ForceUpdate(errors)` ? and check that error callback?

File: internal/server/error.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+type errorsJSONWrapper struct {
+	Errors []string `json:"errors"`
+}
+
+func httpErrors(w http.ResponseWriter, status int, errors []error) {
+	w.WriteHeader(status)
+
+	errs := make([]string, len(errors))
```
Comment (fredericrous) [inline internal/server/error.go:28]: if you make() you need to close() somewhere, right?

File: internal/server/update.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package server
+
+import (
+	"net/http"
+)
+
+func (h *handlers) update(w http.ResponseWriter, r *http.Request) {
+	start := h.timeNow()
+	errors := h.runner.ForceUpdate()
```
Comment (fredericrous) [inline internal/server/update.go:9]: issue here as well

**Review by fredericrous — CHANGES_REQUESTED** [review]

File: internal/server/error.go (hunk 2 above)
Comment (qdm12) [reply internal/server/error.go:28]: Ha that's a fair question! You don't need to close slices (=dynamic arrays in Go) in Go, they just get garbage collected.

You can however close *channels* (like `make(chan string)`) but that's more of a developer thing rather than a performance/memory thing. Someone explained it better than me though:

> You needn't close every channel when you've finished with it. It's only necessary to close a channel when it is important to tell the receiving goroutines that all data have been sent. A channel that the garbage collector determinies to be unreachable will have its resources reclaimed whether or not it is closed. (Don't confuse this with the close operation for open files. It is important to call the Close method on every file when you've finished with it.)

File: cmd/updater/main.go (hunk 1 above)
Comment (qdm12) [reply cmd/updater/main.go]: That was because I forgot to create the `forceResult` channel, it's now fixed 😄 Thanks for finding this out!

And it's now running in a goroutine as we indeed don't want to block the rest of the goroutines with the first round of updates, especially since errors are logged within the updater so we don't need to collect errors.

Comment (fredericrous) [reply cmd/updater/main.go]: hahaha, I should have checked the constructor ^^. I'm glad I could help

File: internal/server/error.go (hunk 2 above)
Comment (fredericrous) [reply internal/server/error.go:28]: thank you for taking the time to explain

**Review by fredericrous — APPROVED** [review]
Tested with success. However, I would expect /update to update the IP no matter what. And that's not the case with the current implementation. Indeed, if the local database contains the same IP already, it won't update the record on the provider website. It's an edge case but if I execute ddns-updater on my laptop. It updates the dns zone to my home's IP. I then stop ddns-updater because my test id done. Hours later I go on my provider's zonedns, the IP is still the one of my laptop and not the one of my NAS which is at my parent's (bless them ❤️). I decide to update the IP with a force update. I send an email to mum, ask her to click on the link http://ddns-updater.test/update and nothing happens. That's an example backed with a little story, for real I have several ways of knowing my IP without external help (other domain, saas app), but I think you get the use case?

I also created a PR here: https://github.com/qdm12/golibs/pull/2 I would love to get debug logs on functions getRecordIDsToUpdate, shouldUpdateRecordWithLookup, shouldUpdateRecordNoLookup

PS: favicon issue is fixed!

**Comment by qdm12** [comment]
> Indeed, if the local database contains the same IP already, it won't update the record on the provider website

That's only the case for cloudflare if the record is proxied. Otherwise it DNS resolves your hostname and check it matches your current public IP address. [This](https://github.com/qdm12/ddns-updater/blob/master/internal/update/run.go#L146) is the codepath ran when there is an update call (through http or with the timer). Are you sure it doesn't update it?

Let's merge this PR for now, as well as the favicon one, but let's continue the discussion here 👍 

Regarding golibs, thanks for that contribution, it's merged now! And yes adding some debug logs could be nice.

On a side note, I'm actually reworking the logger implementation (maybe interface) to have something [like this](https://github.com/WireGuard/wireguard-go/blob/master/device/logger.go#L36) and ideally have different implementations such as zap & glog.

**Comment by fredericrous** [comment]
Yeah it didn't get updated in my case. I had to update the IP manually. In the codepath you linked, IP always seems to equal lastIP in my case. Indeed my home's IP didn't change, what changed was the IP set on the provider. I'll add some debug logs now that is possible ^^ and send a report

Having an interface on top of logging third party is indeed a good thing. A few years ago I wrote a small tool in NodeJS that used Winston logger. I replaced it with Bunyan. I was proud because my tool had a noticeable speed increase. Then I got some user feedback complaining about Bunyan requiring Python. People who had not configured npm and proxies correctly wouldn't be able to download the binary, as a fallback npm was trying to compile the package and this would fail for these users... I had to rollback because I didn't write an interface and had no more time to spend on the tool..

**Comment by qdm12** [comment]
I'm still working on the logging library with a few implementations, I should be done this weekend, and will update golibs in ddns-updater so that you can add debug logs where you please 😉

**Comment by qdm12** [comment]
@fredericrous Have a look at https://github.com/qdm12/golibs/pull/3 about logging please 😉

Any feedback / suggestions / critics are welcome! I've been toying with it all my Saturday afternoon pretty much 😅
